op-node-derivers.md

```diff
@@ -0,0 +1,554 @@
+there should be an explanation of why one solution was picked over other solutions.
+As a rule of thumb, including code snippets (except for defining an external API) is likely too low level.* -->
+
+## Desired properties
+
+To sum up what did not work in above alternative solutions, and what we aim to achieve with the proposed solution:
+
+- Ability to process effects in parallel, grouped by resource, without enshrining the resource.
+- Ability to merge various triggers that may affect the same group of resource, to save on resources.
+- Ability to prioritize effects, if multiple apply
+  (a resource may have not picked up on subsequent signals if it was busy).
+- Ability to bundle conditions/effects, to not require sparse changes from rollup-node mods.
+- Ability to rewire existing effects without rewriting or in-place changes, to make code more reusable.
+  E.g. interop inserts an intermediate safety level.
+- Ability to compose the stack: composition, not object-oriented but extensible, fits most languages and reduces code.
+- Ability to couple effects: changes often cannot be atomically coupled like a DB,
+  but intermediate failures should be possible to respond to, to maintain some level of consistency.
+- Ability to transfer effects between threads without dead-locking or blocking multiple resources.
+
+## Solution
+
+### Derivers and categorized events
+
+By categorizing events, we can schedule processing in parallel, without enshrining specific resources:
+```go
+type Resource string
+
+const (
+    UnsafeL2 Resource = "unsafe-l2"
+    PromoteUnsafeL2 Resource = "promote-unsafe-l2" // example: to not tightly couple "unsafe" to "safe" transition
+    SafeL2 Resource = "safe-l2"
+    FinalizedL2 Resource = "finalized-l2"
+    CurrentL1 Resource = "current-l1" // traversal point on L1 for block-attributes generation
+  
+    // ... standard resources can be extended by mods / new projects, e.g.:
+    CrossUnsafeL2 Resource = "cross-unsafe-l2"
+    
+    // some functionality may gets its own category, to synchronize the work into one place, for less lock contention
+    EngineL2 Resource = "engine-l2"
+)
+
+type Event struct {
+    Resource Resource
+    Data any
+}
+```
+
+We can introduce a `Deriver` as an event processor:
+```go
+type Deriver interface {
+    OnEvent(ctx context.Context, event *Event)
+}
+```
+A deriver is assumed to read-access to any state it needs. Composed by the constructor of the deriver.
+
+#### Composable derivers
+
+Writes idiomatically happen by implementing a purpose-specific Deriver:
+```go
+type Engine struct {
+  control EngineController
+}
+
+func (e *Engine) OnEvent(ctx context.Context, event *Event) {
+    switch x := event.Data.(type) {
+        case *NewPayload:
+            select {
+              // events can include ways to communicate the result, if necessary, across threads
+              case x.Result <- e.onNewPayload(x.Envelope):
```

Comment:
This may break the ability for fault proofs to run everything synchronously - particularly if that channel might fill up and block.